### DISCUSSION TITLE
Those who need must-run constraints using PyPSA-Earth

### DISCUSSION THREAD

**Discussion by RogerKwak** [pypsa-meets-earth/pypsa-earth#1224] [Ideas]
To users of PyPSA-Earth who need must-run constraints in their models:

Recently, I found that PyPSA-Earth lacks must-run constraints, which made it difficult to achieve realistic dispatch patterns.

In PyPSA-Eur, constraints such as 'conventional:  unit_commitment: True/false' can be applied via the unit_commitment.csv file, allowing for dispatch constraints.

However, in the case of PyPSA-Earth, I added must-run constraints for each carrier by modifying the solve_network.py script. I would like to share this with those who may not be aware.

By applying this constraint, it is possible to incorporate base-load generation and must-run conditions into your model.

That said, since this is a forced generation constraint, careful consideration is needed when setting the range. If the range is too narrow, the model may become infeasible.

Ultimately, while adding constraints such as ramp-up, ramp-down, and start-up costs to PyPSA-Earth remains a challenge, I hope this workaround can help improve your modeling.

   ```
  def prepare_network(n, solve_opts):
    if "clip_p_max_pu" in solve_opts:
        for df in (n.generators_t.p_max_pu, n.storage_units_t.inflow):
            df.where(df > solve_opts["clip_p_max_pu"], other=0.0, inplace=True)
    load_shedding = solve_opts.get("load_shedding")
    if load_shedding:
        n.add("Carrier", "Load")
        buses_i = n.buses.query("carrier == 'AC'").index
        if not np.isscalar(load_shedding):
            load_shedding = 5e3  # Eur/kWh
        # intersect between macroeconomic and surveybased
        # 1e2 is practical relevant, 8e3 good for debugging
        n.madd(
            "Generator",
            buses_i,
            " load",
            bus=buses_i,
            carrier="load",
            sign=1e-3,  # Adjust sign to measure p and p_nom in kW instead of MW
            marginal_cost=load_shedding,
            p_nom=1e9,  # kW
        )
    if solve_opts.get("noisy_costs"):
        for t in n.iterate_components(n.one_port_components):
            # TODO: uncomment out to and test noisy_cost (makes solution unique)
            # if 'capital_cost' in t.df:
            #    t.df['capital_cost'] += 1e1 + 2.*(np.random.random(len(t.df)) - 0.5)
            if "marginal_cost" in t.df:
                t.df["marginal_cost"] += 1e-2 + 2e-3 * (
                    np.random.random(len(t.df)) - 0.5
                )
        for t in n.iterate_components(["Line", "Link"]):
            t.df["capital_cost"] += (
                1e-1 + 2e-2 * (np.random.random(len(t.df)) - 0.5)
            ) * t.df["length"]

    if solve_opts.get("nhours"):
        nhours = solve_opts["nhours"]
        n.set_snapshots(n.snapshots[:nhours])
        n.snapshot_weightings[:] = 8760.0 / nhours

     # Set nuclear generators to dispatch between 40% and 80% of their capacity
    nuclear_generators = n.generators.query("carrier == 'nuclear'").index
    n.generators_t.p_min_pu.loc[:, nuclear_generators] = 0.4
    n.generators_t.p_max_pu.loc[:, nuclear_generators] = 0.8

    # Set coal generators to dispatch between 20% and 80% of their capacity
    coal_generators = n.generators.query("carrier == 'coal'").index
    n.generators_t.p_min_pu.loc[:, coal_generators] = 0.2
    n.generators_t.p_max_pu.loc[:, coal_generators] = 0.8

    return n
```

**Comment by ekatef** [comment]
Hello @RogerKwak, and thank you so much for sharing your experience. Completely support your perspective on the need to build scenarios which are aligned with the existing national policies and operational practices. May you be interested to open a PR to add an option for `must_run` into the code?

**Reply by RogerKwak** [reply]
I wish to.

I'm working on the constraints options like PyPSA-Eur below.
'''conventional: 
unit_commitment: '''

I've tried hundreds times to adjust ramp_start/shut down, min_up/down time, but it doesn't work well.

From what I see, it seems like the structure needs to be changed.

Where and what should I change..

**Reply by ekatef** [reply]
> I wish to.
> 
> I'm working on the constraints options like PyPSA-Eur below. '''conventional: unit_commitment: '''
> 
> I've tried hundreds times to adjust ramp_start/shut down, min_up/down time, but it doesn't work well.
> 
> From what I see, it seems like the structure needs to be changed.
> 
> Where and what should I change..

Hello @RogerKwak, apologies have missed to answer here.

As discussed, [this PR](https://github.com/PyPSA/PyPSA/pull/1007) should be merged to make it possible combine expansion studies with unit commitment. Currently, those options are not compatible.

So, the simplified solution you have found is corresponds to state-of-the-art for now. If you wish feel free to contribute your implementation of `must-run` constraint creating [a pull request.](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork). My feeling is that you have still the most part of implementation ready.  The only part left is to add to a configuration file the parameters needed to quantify the `must_run` constraint (a name of technology, as well as upper and lower limits). If you have time to work on it, happy to support.
